### ISSUE TITLE
Shortcut syntax for tuples with simple constructors

### ISSUE THREAD

**Issue by timor** [factor/factor#2451]
I find myself writing this pattern quite frequently:
```factor
TUPLE: descriptive-long-tuple-name ... ;
C: <descriptive-long-tuple-name> descriptive-long-tuple-name
```
That is a bit redundant, since what I want to express is "Make a tuple and give me the default constructor".

Would it make sense to have something like 
```factor
TUPLE: descriptive-long-tuple-name ... ; boa-constructor
```
or

```factor
TUPLE<>: descriptive-long-tuple-name ... ;
```

for these cases?

**Comment by mrjbq7** [comment]
we auto-generate an instance? word. Maybe we should auto generate a boa
constructor.

**Comment by mrjbq7** [comment]
Technically you can just “classname boa”, without the constructor word!

**Comment by timor** [comment]
> Technically you can just “classname boa”, without the constructor word!

True, but in my experience that makes refactoring harder when you decide to change the constructor implementation later, so I try to avoid it.

**Comment by timor** [comment]
> we auto-generate an instance? word. Maybe we should auto generate a boa constructor.

That's a lot of extra words... maybe they can kinda be auto-generated when used first?

**Comment by mrjbq7** [comment]
One of the reasons we have the C: syntax is to have less auto-generate, less magic. 

I kind of want to remove the “foo?” words and instead replace with “foo instance?”.

**Comment by jonenst** [comment]
Auto generating boa constructors hurts greppability.. doesn't mean it's
nécessaire a idea though.

Le mar. 6 avr. 2021 à 17:02, John Benediktsson ***@***.***> a
écrit :

> One of the reasons we have the C: syntax is to have less auto-generate,
> less magic.
>
> I kind of want to remove the “foo?” words and instead replace with “foo
> instance?”.
>
>
>

**Comment by timor** [comment]
On the other hand, if they turn up while grepping, you know that they have been overwritten.  But I still think it is better to generate them only on demand.

**Comment by mrjbq7** [comment]
The solution is obviously to use ndstn.

(non-descriptive-short-tuple-names).

**Comment by timor** [comment]
> > Technically you can just “classname boa”, without the constructor word!
> 
> True, but in my experience that makes refactoring harder when you decide to change the constructor implementation later, so I try to avoid it.

This problem could be solved with generic constructors (class methods), since then you would simply supply a method on your class that specializes `boa` (CLOS kinda like does this).

**Comment by erg** [comment]
It used to just be ``C: <foo>`` but I thought it was too magical, so now you can grep for ``foo`` and find ``C: <foo> foo``. There's also ``CONSTRUCTOR:`` which has more features than plain ``boa`` like not breaking when you change the tuple slot arity and letting you initialize slots in the definition body. It would be cool to get rid of ``C:`` somehow.

We also have the pattern where ``: <foo-subclass> ( class a b c -- obj ) new swap >>c swap >>b swap >>a ;``. Maybe that needs its own syntax? Basically if you want to clean this up, find all the ``<foo>`` words and see which patterns you can abstract out.

**Comment by timor** [comment]
> We also have the pattern where `: <foo-subclass> ( class a b c -- obj ) new swap >>c swap >>b swap >>a ;`. Maybe that needs its own syntax? Basically if you want to clean this up, find all the `<foo>` words and see which patterns you can abstract out.

That may be outside the scope of this PR. The pattern I intended to shorten is the one with default constructors only. 

I actually think that the correct abstraction for `: <foo-subclass> ( a b c class -- obj ) new swap >>c swap >>b swap >>a ;` involves class methods.
Example (`CM:` defines class methods):
```factor
IN: scratchpad auto-use TUPLE: foo a b c ;
IN: scratchpad auto-use TUPLE: foosub < foo ;
IN: scratchpad auto-use GENERIC: <foo-subclass> ( a b c class -- obj )
IN: scratchpad auto-use CM: foo <foo-subclass> new swap >>c swap >>b swap >>a ;
IN: scratchpad auto-use 1 2 3 foo <foo-subclass> .
T{ foo { a 1 } { b 2 } { c 3 } }
IN: scratchpad auto-use 1 2 3 foosub <foo-subclass> .
T{ foosub { a 1 } { b 2 } { c 3 } }
```
That is of course much more than a syntactical abstraction.  However, numerous times when I was using parameterized constructors, this is what I ended up emulating.
